[Webhook] Add new component documentation (#17964)
symfony/symfony-docs: webhook.rst

Webhook
=======

.. versionadded:: 6.3

>     The Webhook component was introduced in Symfony 6.3.

Reviewer: I think we should indicate component is experimental

The Webhook component aims at easing the setup of callback functions
through HTTP in your application.

Webhooks allow third-party services to communicate with your application
without you having to make a request to the third-party application.
By providing a webhook URL to the remote service, it will be able to
send a request to this URL when a predefined event occurs.

> Unlike an API, it is the third party service that defines the payload that

Reviewer: ```suggestion
Unlike an API, it is the third-party service that defines the payload that
```

will be sent to your own endpoint. It is therefore up to you to adapt to
the service.

A recurrent use case is the return of the mail provider.
Some mail providers set up a service allowing you to retrieve the status
of one or more emails you have sent and thus know if they have been
delivered or not. These services then use your webhook to send you this
information once they have it, so that you can store this information
and process it on your side.

Installation
------------

You can install the Webhook component with:

.. code-block:: terminal

    $ composer require symfony/webhook

> 

Reviewer: ```suggestion

.. include:: /components/require_autoload.rst.inc

```

        public function __construct(string $mailId, array $payload, private readonly bool $delivered)
        {
            parent::__construct(self::EVENT_NAME, $mailId, $payload);
        }

        public function isDelivered(): bool
        {
            return $this->delivered;
        }
    }

    // src/RemoteEvent/MailerProviderPayloadConverter.php
    namespace App\RemoteEvent;

    use Symfony\Component\RemoteEvent\Exception\ParseException;
    use Symfony\Component\RemoteEvent\PayloadConverterInterface;

    // This converter transforms a provider specific payload to
    // our generic mailer remote event
    class SpecificMailerProviderPayloadConverter implements PayloadConverterInterface
    {
        public function convert(array $payload): MailerRemoteEvent
        {
            // Payload contains all the information your email provider sends you
            if (!isset($payload['mail_uid'])) {
                throw new ParseException('This payload must contain a mail uid.');
            }

            return new MailerRemoteEvent($payload['mail_uid'], $payload, $payload['delivered'] ?? false);
        }
    }

> The ``MailerProviderPayloadConverter`` can now be injected in our request parser

Reviewer: ```suggestion
The ``SpecificMailerProviderPayloadConverter`` can now be injected in our request parser
```

        ) {
        }

        protected function getRequestMatcher(): RequestMatcherInterface
        {
            return new ChainRequestMatcher([
                new MethodRequestMatcher('POST'),
                new IsJsonRequestMatcher(),
            ]);
        }

        protected function doParse(Request $request, string $secret): ?RemoteEvent
        {
            $content = $request->toArray();

            try {
                // Use the new converter to create our custom remote event
                return $this->converter->convert($content);
            } catch (ParseException $e) {
                throw new RejectWebhookException(406, $e->getMessage(), $e);
            }
        }
    }

Validate Request Preconditions
------------------------------

By using the :class:`Symfony\\Component\\HttpFoundation\\ChainRequestMatcher`,
it is possible to build a powerful request validation chain to determine
if the request that arrived in your webhook endpoint is valid and can
be processed. The following matchers are available:

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\AttributesRequestMatcher`

>     Matches request's attributes against regex

Reviewer 2: I don't understand this well. Maybe we could add a short example about how to define that regex? Thanks!

Author: That's updated with an example. Is it better?

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\ExpressionRequestMatcher`
    Matches the request against an expression using the ExpressionLanguage component

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\HostRequestMatcher`
    Matches the host that sent the request

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\IpsRequestMatcher`
    Matches an IP or a set of IPs from where the request comes from

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\MethodRequestMatcher`

>     Matches the request uses a given method

Reviewer 2: `a given method` -> `a given HTTP method`

Author: Updated as well!

Webhook
=======

.. versionadded:: 6.3

    The Webhook component was introduced in Symfony 6.3 and is marked
    as experimental.

The Webhook component aims at easing the setup of callback functions
through HTTP in your application.

Webhooks allow third-party services to communicate with your application
without you having to make a request to the third-party application.
By providing a webhook URL to the remote service, it will be able to

> send a request to this URL when a predefined event occurs.

Reviewer 3: ```suggestion
send a request to your URL when a predefined event occurs in their system.
```

Unlike an API, it is the third-party service that defines the payload that
will be sent to your own endpoint. It is therefore up to you to adapt to

> the service.

Reviewer 3: ```suggestion
the service to handle such payload accordingly.
```

> A recurrent use case is the return of the mail provider.

Reviewer 3: As it may not be that recurrent, what about just add instead: As an example use case, email providers can send webhook…

Some mail providers set up a service allowing you to retrieve the status
of one or more emails you have sent and thus know if they have been
delivered or not. These services then use your webhook to send you this
information once they have it, so that you can store this information
and process it on your side.

Installation
------------

You can install the Webhook component with:

.. code-block:: terminal

    $ composer require symfony/webhook

.. include:: /components/require_autoload.rst.inc

Basic Usage
-----------

Parse The Incoming Request
~~~~~~~~~~~~~~~~~~~~~~~~~~

> Each type of webhook has its own parser. A parser is a service that has

Reviewer 3: ```suggestion
Each type of webhook has its own parser. A parser is a service class that has
```

perhaps redondant (service is a class) but commenting just

the ability to verify that a request meets certain preconditions to be

> processed. Here is an example of a basic parser::

Reviewer 3: ```suggestion
processed and return an event to be handles by your application. Here is an example of a basic parser::
```

            return new RemoteEvent('mailer_callback.event', 'event-id', $content);
        }
    }

.. tip::

    If you need more flexibility, you can also create your parser by
    implementing the
    :class:`Symfony\\Component\\Webhook\\Client\\RequestParserInterface`
    interface.

If the webhook request matches all preconditions, the ``doParse()`` method is executed
and returns a :class:`Symfony\\Component\\RemoteEvent\\RemoteEvent`.

Consume Remote Events
~~~~~~~~~~~~~~~~~~~~~

Once a remote event has been created, it has to be consumed by a **remote event consumer**.
A remote event consumer can be declared in two ways: by implementing
the :class:`Symfony\\Component\\RemoteEvent\\Consumer\\ConsumerInterface`,
or by using the
:class:`Symfony\\Component\\RemoteEvent\\Attribute\\AsRemoteEventConsumer` attribute.
In both case, the consumer must implement the ``consume()`` method::

    use Symfony\Component\RemoteEvent\Attribute\AsRemoteEventConsumer;
    use Symfony\Component\RemoteEvent\RemoteEvent;

    #[AsRemoteEventConsumer(name: 'mailer_callback.event')]
    class MailerCallbackEventConsumer
    {
        public function consume(RemoteEvent $event): void
        {

>             // Process the event returned by our parser

Reviewer 3: ```suggestion
            // Process the event returned by your parser
```

as above you document: « create your parser«

    Matches a request's attributes against a regex. For example, if your request
    URL looks like ``/users/{name}``, you may want to check that the ``name``
    attribute only contains alphanumeric characters with a regex like
    ``[a-zA-Z0-9]+``.

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\ExpressionRequestMatcher`
    Matches the request against an expression using the ExpressionLanguage component.

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\HostRequestMatcher`
    Matches the host that sent the request.

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\IpsRequestMatcher`
    Matches an IP or a set of IPs from where the request comes from.

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\MethodRequestMatcher`
    Matches the request uses a given HTTP method.

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\PathRequestMatcher`
    Matches the request path.

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\PortRequestMatcher`
    Matches the request port.

:class:`Symfony\\Component\\HttpFoundation\\RequestMatcher\\SchemeRequestMatcher`
    Matches the request scheme.

> 

Reviewer 3: Missing the IsJsonRequestMatcher is’nt it?

    final class MailerWebhookParser extends AbstractRequestParser
    {
        protected function getRequestMatcher(): RequestMatcherInterface
        {
            return new ChainRequestMatcher([
                new MethodRequestMatcher('POST'),
                new IsJsonRequestMatcher(),
            ]);
        }

        protected function doParse(Request $request, string $secret): ?RemoteEvent
        {
            $content = $request->toArray();
            if (!isset($content['signature']['token'])) {
                throw new RejectWebhookException(406, 'Payload is malformed.');
            }

            return new RemoteEvent('mailer_callback.event', 'event-id', $content);
        }
    }

.. tip::

    If you need more flexibility, you can also create your parser by
    implementing the
    :class:`Symfony\\Component\\Webhook\\Client\\RequestParserInterface`
    interface.

If the webhook request matches all preconditions, the ``doParse()`` method is executed
and returns a :class:`Symfony\\Component\\RemoteEvent\\RemoteEvent`.
This remote event have a unique name. This name will be used to identify which

> consumer has to be invoked to handle it. These consumers are called

Reviewer 4: I'm not sure about it. The right consumer is retrieved by the `{type}` in the webhook path (`/webhook/slack` for instance). The `name` property of a `RemoteEvent` object is set during the handling of an event.

    as experimental.

The Webhook component aims at easing the setup of callback functions
through HTTP in your application.

Webhooks allow third-party services to communicate with your application
without you having to make a request to the third-party application.
By providing a webhook URL to the remote service, it will be able to
send a request to your URL when a predefined event occurs in their system.
Unlike an API, it is the third-party service that defines the payload that
will be sent to your own endpoint. It is therefore up to you to adapt to
the service to handle such payload accordingly.

As an example of use case, some mail providers set up a service allowing
you to retrieve the status of one or more emails you have sent
and thus know if they have been delivered or not.
These services then use your webhook to send you this
information once they have it, so that you can store this information
and process it on your side.

Installation
------------

You can install the Webhook component with:

.. code-block:: terminal

    $ composer require symfony/webhook

.. include:: /components/require_autoload.rst.inc

> Basic Usage

Reviewer 5: I'd add a small 'Configuration' section  after this with a short description, and a small example:
`.. configuration-block::

    .. code-block:: yaml

        # config/packages/framework.yaml
        framework:
            webhook:
                routing:
                    email:
                        service: App\Webhook\MailerWebhookParser # custom parser id
                        secret: '%env(EXTERNAL_SERVICE_SECRET)%'` # this can be validated in the doPrase method

Author: I don't know if we should add this in the component documentation or in the framework documentation? 🤔

Reviewer 5: From what I've seen on other documentation pages, this is included in the component doc, not in the framework doc

Author: Added :+1:
